### DISCUSSION TITLE
process is not defined when using a package that depends on it (in client)

### DISCUSSION THREAD

**Discussion by chang-ryan** [remix-run/remix#4906] [Q&A] [answered]
## Read me!

I fully understand that from several other issues mentioned here:
https://github.com/remix-run/remix/issues/1186, https://github.com/remix-run/history/issues/821, https://github.com/remix-run/remix/issues/2927

We should not be using process.env in the client. However, this issue is not concerning that. It's concerning the fact that other packages might be using it.

Wanted to start a discussion on this.

### What version of Remix are you using?

1.8.2

### Steps to Reproduce

1. install remix indie stack
2. install `@paperxyz/react-client-sdk`
3. use package `import { PaperSDKProvider, LoginWithPaper } from "@paperxyz/react-client-sdk";`
4. `npm run dev`
5. bug occurs

### Expected Behavior

I expect us to be able to pull in this package `@paperxyz/react-client-sdk`

### Actual Behavior

```
Uncaught ReferenceError: process is not defined
    at node_modules/@paperxyz/js-client-sdk/dist/index.js (index.js:5:1534)
```

This is because the client dependency `@paperxyz/react-client-sdk` has references to `process.env` in its source code. I can monkey patch this by directly going into the outputted `index.js` file from `@paperxyz` and setting the following

```
// inside `node_modules/@paperxyz/js-client-sdk/dist/index.js`

var process = {
  env: {}
};
```

Is there any way I can globally inject this fake variable into remix's client environment? I anticipate that other packages that utilize a reference to `process.env` in its source code will break with remix.

**Comment by buildtheui** [comment]
This is also happening to me in another context. We have a monorepo and we're importing some local packages that are used on the client side, these libs like a UI component library or shared configs are using the process.env behind-the-scenes reading variables injected for the whole monorepo, this is working for other apps we have created using other frameworks like Nextjs or simple CRA apps and changing the way Envs are used in the shared libs would break other existing apps, so this is similar to what @chang-ryan mentions, is there any way to solve this?

**Reply by chang-ryan** [reply]
i'm using this as a workaround

```
`window.process = ${JSON.stringify({})}`
```

```
// root.tsx

export default function App() {
  return (
    <html lang="en" className="h-full">
      <head>
        <Meta />
        <Links />
      </head>
      <body className="h-full">
        <div className="absolute w-full top-0 z-20 bg-gradient-to-b from-slate-800 flex justify-between items-center py-4 px-6">
          <img src={nectarverseLogo} alt="Nectarverse logo" className="h-12" />
          <nav className="flex items-center">
            <Link to="" className="font-bold uppercase text-white ml-4">Nectar</Link>
            <Link to="" className="font-bold uppercase text-white ml-4">Roadmap</Link>
            <Link to="" className="font-bold uppercase text-white ml-4">Explore</Link>
            <Link to="" className="flex items-center justify-center rounded-md border border-transparent bg-nectar-green/30 px-20 py-1 text-base font-bold uppercase text-white shadow-sm sm:px-8-md ml-12">Home</Link>
          </nav>
        </div>

        <Outlet />
        <ScrollRestoration />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.process = ${JSON.stringify({})}`,
          }}
        />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

**Reply by buildtheui** [reply]
Thanks @chang-ryan this workaround works for serializable values, I wish we wouldn't have to do that, but it's what we have for now

**Comment by sergiodxa** [comment, marked as answer]
You can use an inline script to define `process.env`, see https://sergiodxa.com/articles/use-process-env-client-side-with-remix

**Reply by chang-ryan** [reply]
this is a nice way to inject env variables from `.env` files into the client side application, but if there are dependencies on the front end (from another package that i've imported), then this issue will still cause an error

**Reply by chang-ryan** [reply]
for example, this dependency was breaking for me: https://github.com/paperxyz/js-client-sdk/blob/main/src/constants/settings.ts

but you can see from the code that they've since patched it (at my request) with `?`. however like @buildtheui said, there could be other dependencies that use `process.env` behind the scenes. so it might be fruitful to have some kind of workaround in remix for this. otherwise we need to just wait on the dependency maintainers to patch any usages of `process.env`

**Reply by jdnichollsc** [reply]
we're not able to use objects from _process.env_ like **Buffer** with other dependencies like web3 packages, etc using inline scripts

**Reply by buildtheui** [reply]
@jdnichollsc correct for non-serializable values we have to resort to other things like patching the compiler as @jdnichollsc did, what I ended up doing in this particular case was injecting the Buffer class in a useEffect on the root class as follows:
```js
// root.tsx
import { Buffer } from "buffer";
...
useEffect(() => {
        // This allows WalletConnect to work
        if (!window.Buffer) {
            window.Buffer = Buffer;
        }
 });
...
```

**Reply by jdnichollsc** [reply]
@buildtheui sorry for the delay, I tried that alternative but I was getting this error:
```
index.js:6 Uncaught ReferenceError: process is not defined
    at ../../node_modules/@toruslabs/openlogin-jrpc/node_modules/pump/index.js (index.js:6:29)
    at __require2 (chunk-PZDJHGND.js:18:50)
    at slicedToArray.js:7:1
```

**Comment by jdnichollsc** [comment]
I was able to solve issues with libs (web3auth, etc) using **process.env** and **Buffer** from _Browser_ with the below patch:

- patches/@remix-run+dev+1.10.1.patch
```.patch
diff --git a/node_modules/@remix-run/dev/dist/compiler/compileBrowser.js b/node_modules/@remix-run/dev/dist/compiler/compileBrowser.js
index 1fa83a5..5892640 100644
--- a/node_modules/@remix-run/dev/dist/compiler/compileBrowser.js
+++ b/node_modules/@remix-run/dev/dist/compiler/compileBrowser.js
@@ -16,11 +16,11 @@ var path = require('path');
 var module$1 = require('module');
 var esbuild = require('esbuild');
 var nodeModulesPolyfill = require('@esbuild-plugins/node-modules-polyfill');
+var nodeGlobalsPolyfill = require('@esbuild-plugins/node-globals-polyfill');
 var assets = require('./assets.js');
 var dependencies = require('./dependencies.js');
 var loaders = require('./loaders.js');
 var browserRouteModulesPlugin = require('./plugins/browserRouteModulesPlugin.js');
-var cssFilePlugin = require('./plugins/cssFilePlugin.js');
 var deprecatedRemixPackagePlugin = require('./plugins/deprecatedRemixPackagePlugin.js');
 var emptyModulesPlugin = require('./plugins/emptyModulesPlugin.js');
 var mdx = require('./plugins/mdx.js');
@@ -77,10 +77,18 @@ const createEsbuildConfig = (config, options) => {
     // that we don't want to run in the browser (i.e. action & loader).
     entryPoints[id] = config.routes[id].file + "?browser";
   }
-  let plugins = [deprecatedRemixPackagePlugin.deprecatedRemixPackagePlugin(options.onWarning), cssFilePlugin.cssFilePlugin({
-    mode: options.mode,
-    rootDirectory: config.rootDirectory
-  }), urlImportsPlugin.urlImportsPlugin(), mdx.mdxPlugin(config), browserRouteModulesPlugin.browserRouteModulesPlugin(config, /\?browser$/), emptyModulesPlugin.emptyModulesPlugin(config, /\.server(\.[jt]sx?)?$/), nodeModulesPolyfill.NodeModulesPolyfillPlugin()];
+  let plugins = [
+    nodeGlobalsPolyfill.NodeGlobalsPolyfillPlugin({
+      buffer: true,
+      process: true
+    }),
+    nodeModulesPolyfill.NodeModulesPolyfillPlugin(),
+    deprecatedRemixPackagePlugin.deprecatedRemixPackagePlugin(options.onWarning),
+    urlImportsPlugin.urlImportsPlugin(),
+    mdx.mdxPlugin(config),
+    browserRouteModulesPlugin.browserRouteModulesPlugin(config, /\?browser$/),
+    emptyModulesPlugin.emptyModulesPlugin(config, /\.server(\.[jt]sx?)?$/)
+  ];
   return {
     entryPoints,
     outdir: config.assetsBuildDirectory,
diff --git a/node_modules/@remix-run/dev/dist/compiler/compilerServer.js b/node_modules/@remix-run/dev/dist/compiler/compilerServer.js
index 21cbad7..d70b0e3 100644
--- a/node_modules/@remix-run/dev/dist/compiler/compilerServer.js
+++ b/node_modules/@remix-run/dev/dist/compiler/compilerServer.js
@@ -16,8 +16,8 @@ var path = require('path');
 var esbuild = require('esbuild');
 var fse = require('fs-extra');
 var nodeModulesPolyfill = require('@esbuild-plugins/node-modules-polyfill');
+var nodeGlobalsPolyfill = require('@esbuild-plugins/node-globals-polyfill');
 var loaders = require('./loaders.js');
-var cssFilePlugin = require('./plugins/cssFilePlugin.js');
 var deprecatedRemixPackagePlugin = require('./plugins/deprecatedRemixPackagePlugin.js');
 var emptyModulesPlugin = require('./plugins/emptyModulesPlugin.js');
 var mdx = require('./plugins/mdx.js');
@@ -63,11 +63,14 @@ const createEsbuildConfig = (config, assetsManifestChannel, options) => {
   }
   let isCloudflareRuntime = ["cloudflare-pages", "cloudflare-workers"].includes(config.serverBuildTarget ?? "");
   let isDenoRuntime = config.serverBuildTarget === "deno";
-  let plugins = [deprecatedRemixPackagePlugin.deprecatedRemixPackagePlugin(options.onWarning), cssFilePlugin.cssFilePlugin({
-    mode: options.mode,
-    rootDirectory: config.rootDirectory
-  }), urlImportsPlugin.urlImportsPlugin(), mdx.mdxPlugin(config), emptyModulesPlugin.emptyModulesPlugin(config, /\.client(\.[jt]sx?)?$/), serverRouteModulesPlugin.serverRouteModulesPlugin(config), serverEntryModulePlugin.serverEntryModulePlugin(config), serverAssetsManifestPlugin.serverAssetsManifestPlugin(assetsManifestChannel.read()), serverBareModulesPlugin.serverBareModulesPlugin(config, options.onWarning)];
+  let plugins = [deprecatedRemixPackagePlugin.deprecatedRemixPackagePlugin(options.onWarning), urlImportsPlugin.urlImportsPlugin(), mdx.mdxPlugin(config), emptyModulesPlugin.emptyModulesPlugin(config, /\.client(\.[jt]sx?)?$/), serverRouteModulesPlugin.serverRouteModulesPlugin(config), serverEntryModulePlugin.serverEntryModulePlugin(config), serverAssetsManifestPlugin.serverAssetsManifestPlugin(assetsManifestChannel.read()), serverBareModulesPlugin.serverBareModulesPlugin(config, options.onWarning)];
   if (config.serverPlatform !== "node") {
+    plugins.unshift(
+      nodeGlobalsPolyfill.NodeGlobalsPolyfillPlugin({
+        buffer: true,
+        process: true
+      })
+    );
     plugins.unshift(nodeModulesPolyfill.NodeModulesPolyfillPlugin());
   }
   return {
```

And from package.json:
```.json
  "scripts": {
    "postinstall": "run-s postinstall:*",
    "postinstall:patch": "patch-package",
    "postinstall:remix": "remix setup node",
  },
```

Let me know if there's any solution to avoid disabling cssFilePlugin!

**Reply by jdnichollsc** [reply]
cssFilePlugin error was fixed in the latest version of Remix! <3

**Comment by anakinsolo** [comment]
There is this package https://www.npmjs.com/package/@storyblok/react that you must init inside the root.tsx and outside of anything
```
//app/root.tsx
...
import { apiPlugin, storyblokInit } from '@storyblok/react';
import { ComponentList } from './components/ComponentList';

storyblokInit({
  accessToken: 'token',
  use: [apiPlugin],
  components: ComponentList,
});
...
//The rest of the root file
const MainStyles = ( {children} :{children: ReactNode}) => {
  return (
    <html lang="en" className="h-full">
      <head>
        <Meta />
        <Links />
      </head>
      <body className="h-full">
        {children}
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
};

export default function App() {
  return (
    <MainStyles>
      <Navigation />
      <Outlet />
      <Footer />
    </MainStyles>
  );
}
```
So for me, none of the solutions listed in here works https://remix.run/docs/en/1.14.0/guides/envvars#browser-environment-variables

**Comment by johannbuscail** [comment]
I'm facing the same issue. I need Shopify Admin API to run on a hydrogen app (using remix). But this admin API requires access to the `process` variable. This throws the error `Reference Error: Process is not defined`. Can anyone help ?

**Comment by jdnichollsc** [comment]
Hey folks, check my example to fix this issue https://gist.github.com/jdnichollsc/b73eca1f34bdd1e3a95283774fda2212

**Comment by gustavopch** [comment]
In my case, I was importing utils from outside the app directory. It turns out undesired code ended up in the client bundle because `.server.ts` extension is only considered for files inside it. Fixed by patching the condition below to use `config.rootDirectory` instead of `config.appDirectory`.
https://github.com/remix-run/remix/blob/c9648c7d76c08224af0a0327109878847db41a2b/packages/remix-dev/compiler/plugins/emptyModules.ts#L19-L27

**Comment by zhanybekovv** [comment]
Any updates here? @chang-ryan @sergiodxa @jdnichollsc @anakinsolo @buildtheui

**Reply by sergiodxa** [reply]
Updates on what? Remix is most likely not going to polyfill `process` on the client-side build. Packages that are intended to be used client-side should stop depending on things only available on Node.js.

If the package only uses `process.env` it's possible to render an inline script to set `window.process = { env: {} }` as I explained here https://sergiodxa.com/articles/use-process-env-client-side-with-remix, this way you can set what env variables the package needs.

For other built-ins available only on Node.js, you will have to manually polyfill them, or ask the package author to don't rely on them being available.

Also, please don't mention everyone who commented on a discussion.

**Reply by saugion** [reply]
I'm using a library that has a reference to pump, that is what is causing the process issue. I've solved it with this vite configuration, hope it helps

```
import {UserConfigFn, splitVendorChunkPlugin} from 'vite';
import { overrideVaadinConfig } from './vite.generated';
import { NodeGlobalsPolyfillPlugin } from '@esbuild-plugins/node-globals-polyfill'
import { NodeModulesPolyfillPlugin } from '@esbuild-plugins/node-modules-polyfill'
import rollupNodePolyFill from 'rollup-plugin-polyfill-node'

const customConfig: UserConfigFn = (env) => ({
  // Here you can add custom Vite parameters
  // https://vitejs.dev/config/
    optimizeDeps: {
        esbuildOptions: {
            target: "esnext",
            define: {
                global: 'globalThis'
            },
            plugins: [
                NodeGlobalsPolyfillPlugin({
                    buffer: true,
                    process: true,
                }),
                NodeModulesPolyfillPlugin()
            ]
        },
    },

    build: {
        minify: false,
        target: "esnext",
        commonjsOptions: {
            transformMixedEsModules: true
        },
        rollupOptions: {
            plugins: [
                rollupNodePolyFill()
            ]
        }
    },

    plugins: [splitVendorChunkPlugin()],

    define: {
        "process.env.NODE_ENV": JSON.stringify(env.mode)
    }
});

export default overrideVaadinConfig(customConfig);
```
